File: src/main/java/com/projectswg/holocore/resources/support/objects/swg/creature/CreatureObjectAwareness.java
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/

package com.projectswg.holocore.resources.support.objects.swg.creature;

import com.projectswg.common.data.location.Location;
import com.projectswg.common.network.packets.SWGPacket;
import com.projectswg.common.network.packets.swg.zone.*;
import com.projectswg.common.network.packets.swg.zone.baselines.Baseline.BaselineType;
import com.projectswg.common.network.packets.swg.zone.building.UpdateCellPermissionMessage;
import com.projectswg.common.network.packets.swg.zone.object_controller.DataTransform;
import com.projectswg.common.network.packets.swg.zone.object_controller.DataTransformWithParent;
import com.projectswg.holocore.resources.support.global.player.Player;
import com.projectswg.holocore.resources.support.objects.swg.SWGObject;
import com.projectswg.holocore.resources.support.objects.swg.building.BuildingObject;
import com.projectswg.holocore.resources.support.objects.swg.cell.CellObject;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.concurrent.atomic.AtomicReference;

public class CreatureObjectAwareness {
	
	private final CreatureObject creature;
	private final Set<SWGObject> aware;
	private final Set<Long> awareIds;
	private final Comparator<SWGObject> objectComparator;
	private final AtomicReference<SWGPacket> finalTeleportPacket;
	
	public CreatureObjectAwareness(CreatureObject creature) {
		this.creature = creature;
		this.aware = new HashSet<>();
		this.awareIds = new HashSet<>();
		this.objectComparator = Comparator.comparingInt(CreatureObjectAwareness::getObjectDepth).thenComparingDouble(this::getDistance);
		this.finalTeleportPacket = new AtomicReference<>(null);
	}
	
	public synchronized void setTeleportDestination(SWGObject parent, Location location) {
		if (parent == null)
			finalTeleportPacket.set(new DataTransform(creature.getObjectId(), 0, creature.getNextUpdateCount(), location, 0));
		else
			finalTeleportPacket.set(new DataTransformWithParent(creature.getObjectId(), 0, creature.getNextUpdateCount(), parent.getObjectId(), location, 0));
	}
	
	public synchronized void flush(@NotNull Player target) {
		Set<SWGObject> newAware = creature.getAware();
		flushCreates(target, newAware);
		
		SWGPacket finalTeleportPacket = this.finalTeleportPacket.getAndSet(null);
		if (finalTeleportPacket != null)
			creature.sendSelf(finalTeleportPacket);
		
		flushDestroys(target, newAware);
	}
	
	private void flushCreates(@NotNull Player target, Set<SWGObject> newAware) {
		List<SWGObject> create = new ArrayList<>();
		List<SWGObject> added = new ArrayList<>();
		
		// Create Deltas
		for (SWGObject createCandidate : newAware) {
			if (aware.contains(createCandidate))
				continue;
			added.add(createCandidate);
		}
		getCreateList(create, added);
		
		// Create the objects on the client
		LinkedList<SWGObject> createStack = new LinkedList<>();
		for (SWGObject obj : create) {
			if (isBundledObject(obj, obj.getParent()))
				popStackUntil(target, createStack, obj.getParent());
			else
				popStackAll(target, createStack);
			createStack.add(obj);
			createObject(obj, target);
		}
		popStackAll(target, createStack);
		
		// Server awareness update
		for (SWGObject add : create) { // Using "create" here because it's filtered to ensure no crashes
			aware.add(add);
			awareIds.add(add.getObjectId());
			add.addObserver(creature);
		}
		
		// Hope we didn't screw anything up
		assert aware.contains(creature.getSlottedObject("ghost")) : "not aware of ghost " + creature;
		assert aware.contains(creature) : "not aware of creature";
	}
	
	private void flushDestroys(@NotNull Player target, Set<SWGObject> newAware) {
		List<SWGObject> destroy = new ArrayList<>();
		List<SWGObject> removed = new ArrayList<>();
		
		// Create Deltas
		for (SWGObject destroyCandidate : aware) {
			if (newAware.contains(destroyCandidate))
				continue;
			removed.add(destroyCandidate);
		}
		getDestroyList(destroy, removed);
		
		// Remove destroyed objects so that nobody tries to send a packet to us after we send the destroy
		for (Iterator<SWGObject> it = aware.iterator(); it.hasNext(); ) {
			SWGObject currentAware = it.next();
			for (SWGObject remove : destroy) { // Since the "create" is filtered, aware could also have been filtered
				if (isParent(currentAware, remove)) {
					it.remove();
					awareIds.remove(currentAware.getObjectId());
					remove.removeObserver(creature);
					break;
				}
			}
		}
		
		// Destroy the objects on the client
		for (SWGObject obj : destroy) {
			destroyObject(obj, target);
		}
		
		// Hope we didn't screw anything up
		assert aware.contains(creature.getSlottedObject("ghost")) : "not aware of ghost " + creature;
		assert aware.contains(creature) : "not aware of creature";
	}
	
	public synchronized void resetObjectsAware() {
		for (SWGObject obj : aware) {
			obj.removeObserver(creature);
		}
		aware.clear();
		awareIds.clear();
	}
	
	public synchronized boolean isAware(long objectId) {
		return awareIds.contains(objectId);
	}
	
	public synchronized boolean isAware(SWGObject obj) {
		return aware.contains(obj);
	}
	
	private void getCreateList(List<SWGObject> list, List<SWGObject> added) {
		added.sort(objectComparator);
		for (SWGObject obj : added) {
			SWGObject parent = obj.getParent();
			if (parent != null && !aware.contains(parent) && !list.contains(parent)) {
				assert !(obj instanceof CellObject);
				continue;
			}
			assert !(obj instanceof BuildingObject) || added.containsAll(obj.getContainedObjects()) : "All cells must be sent with the building";
			if (!isBundledObject(obj, parent) || aware.contains(parent)) {
				list.add(obj);
			} else {
				int parentIndex = list.indexOf(parent);
				assert parentIndex != -1 : "parent isn't added along with child";
				list.add(parentIndex+1, obj);
			}
		}
	}
	
	private void getDestroyList(List<SWGObject> list, List<SWGObject> removed) {
		removed.sort(objectComparator);
		for (SWGObject obj : removed) {
			// Don't delete our own parent nor child objects if we're deleting their parent (optimization)
			if (isParent(obj) || list.contains(obj.getParent()))
				continue;
			list.add(obj);
		}
	}
	
	private boolean isParent(SWGObject obj) {
		return isParent(creature, obj);
	}
	
	private double getDistance(SWGObject obj) {
		if (obj.getParent() != null)
			return 0;
		
		return creature.getWorldLocation().distanceTo(obj.getLocation());
	}
	
	private boolean isBundledObject(SWGObject obj, SWGObject parent) {
		return parent != null && (obj.getSlotArrangement() == -1 || obj.getBaselineType() == BaselineType.PLAY || parent == creature);
	}
	
	private void createObject(@NotNull SWGObject obj, @NotNull Player target) {
		long id = obj.getObjectId();
		{ // SceneCreateObjectByCrc
			SceneCreateObjectByCrc create = new SceneCreateObjectByCrc();
			create.setObjectId(id);
			create.setLocation(obj.getLocation());
			create.setObjectCrc(obj.getCrc());
			target.sendPacket(create);
		}
		{ // Baselines
			boolean owner = obj.getOwner() == target;
			
			target.sendPacket(obj.createBaseline3(target));
			target.sendPacket(obj.createBaseline6(target));
			
			if (owner) {
				target.sendPacket(obj.createBaseline1(target));
				target.sendPacket(obj.createBaseline4(target));
				target.sendPacket(obj.createBaseline8(target));
				target.sendPacket(obj.createBaseline9(target));
			}
		}
		{ // Miscellaneous
			if (obj instanceof CellObject)
				target.sendPacket(new UpdateCellPermissionMessage((byte) 1, id));
			// ? UpdatePostureMessage for PlayerObject ?
			if (obj instanceof CreatureObject && obj.isGenerated()) {
				CreatureObject creature = (CreatureObject) obj;
				target.sendPacket(new UpdatePostureMessage(creature.getPosture().getId(), id));
				target.sendPacket(new UpdatePvpStatusMessage(creature.getPvpFaction(), id, this.creature.getPvpFlagsFor(creature)));
			}
		}
		{ // UpdateContainmentMessage
			SWGObject parent = obj.getParent();
			if (parent != null)
				target.sendPacket(new UpdateContainmentMessage(obj.getObjectId(), parent.getObjectId(), obj.getSlotArrangement()));
		}
	}
	
	private static void popStackAll(Player target, LinkedList<SWGObject> createStack) {
		SWGObject parent;
		while ((parent = createStack.pollLast()) != null) {
			target.sendPacket(new SceneEndBaselines(parent.getObjectId()));
		}
	}
	
	private static void popStackUntil(Player target, LinkedList<SWGObject> createStack, SWGObject parent) {
		SWGObject last;
		while ((last = createStack.peekLast()) != null) {
			if (last == parent)
				break;
			createStack.pollLast();
			target.sendPacket(new SceneEndBaselines(last.getObjectId()));
		}
	}
	
	private static void destroyObject(@NotNull SWGObject obj, @NotNull Player target) {
		target.sendPacket(new SceneDestroyObject(obj.getObjectId()));
	}
	
	private static int getObjectDepth(SWGObject obj) {
		return obj == null ? 0 : 1 + getObjectDepth(obj.getParent());
	}
	
	private static boolean isParent(SWGObject child, SWGObject testParent) {
		SWGObject parent = child;
		while (parent != null) {
			if (parent == testParent)
				return true;
			parent = parent.getParent();
		}
		return false;
	}
	
}

File: src/main/java/com/projectswg/holocore/resources/support/objects/swg/creature/CreatureObjectAwareness.kt
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 * *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 * *
 * This file is part of Holocore.                                                  *
 * *
 * --------------------------------------------------------------------------------*
 * *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 * *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 * *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http:></http:>//www.gnu.org/licenses/>.               *
 */

package com.projectswg.holocore.resources.support.objects.swg.creature

import com.projectswg.common.data.location.Location
import com.projectswg.common.network.packets.SWGPacket
import com.projectswg.common.network.packets.swg.zone.*
import com.projectswg.common.network.packets.swg.zone.baselines.Baseline.BaselineType
import com.projectswg.common.network.packets.swg.zone.building.UpdateCellPermissionMessage
import com.projectswg.common.network.packets.swg.zone.object_controller.DataTransform
import com.projectswg.common.network.packets.swg.zone.object_controller.DataTransformWithParent
import com.projectswg.holocore.resources.support.global.player.Player
import com.projectswg.holocore.resources.support.objects.swg.SWGObject
import com.projectswg.holocore.resources.support.objects.swg.cell.CellObject
import com.projectswg.holocore.resources.support.objects.swg.custom.AIObject
import java.util.*
import java.util.concurrent.atomic.AtomicReference

class CreatureObjectAwareness(private val creature: CreatureObject) {
	
	private val aware = HashSet<SWGObject>()
	private val awareIds = HashSet<Long>()
	private val finalTeleportPacket = AtomicReference<SWGPacket>(null)
	private val flushAwarenessData = FlushAwarenessData(creature)
	
	@Synchronized
	fun setTeleportDestination(parent: SWGObject?, location: Location) {
		if (parent == null)
			finalTeleportPacket.set(DataTransform(creature.objectId, 0, creature.nextUpdateCount, location, 0f))
		else
			finalTeleportPacket.set(DataTransformWithParent(creature.objectId, 0, creature.nextUpdateCount, parent.objectId, location, 0f))
	}
	
	@Synchronized
	fun flushNoPlayer() {
		handleFlush({}, {}, {}) // No action required for NPCs
	}
	
	@Synchronized
	fun flush(target: Player) {
		handleFlush(
			createHandler = { create ->
				// Send all SceneCreateObject's and Baselines
				val createStack = LinkedList<SWGObject>()
				for (obj in create) {
					val parent = obj.parent
					if (parent != null) {
						popStackUntil(target, createStack, parent)
						if (!obj.isBundledWithin(parent, creature) && createStack.peekLast() == parent) {
							target.sendPacket(SceneEndBaselines(parent.objectId))
							createStack.pollLast()
						}
					} else {
						popStackAll(target, createStack)
					}
					createStack.add(obj)
					createObject(obj, target)
				}
				popStackAll(target, createStack)
			},
			intermediateCallback = {
				// Perform a teleport if necessary
				val finalTeleportPacket = this.finalTeleportPacket.getAndSet(null)
				if (finalTeleportPacket != null)
					creature.sendSelf(finalTeleportPacket)
			},
			destroyHandler = { destroy ->
				// Destroy the objects on the client
				for (obj in destroy) {
					destroyObject(obj, target)
				}
			})
	}
	
	private inline fun handleFlush(createHandler: (Collection<SWGObject>) -> Unit, intermediateCallback: () -> Unit, destroyHandler: (Collection<SWGObject>) -> Unit) {
		val newAware = creature.aware
		handleFlushCreate(newAware, createHandler)
		intermediateCallback()
		handleFlushDestroy(newAware, destroyHandler)
	}
	
	private inline fun handleFlushCreate(newAware: Set<SWGObject>, createHandler: (Collection<SWGObject>) -> Unit) {
		val create = flushAwarenessData.buildCreate(aware, newAware)
		createHandler(create)
		
		for (add in create) { // Using "create" here because it's filtered to ensure no crashes
			aware.add(add)
			awareIds.add(add.objectId)
			add.addObserver(creature)
			creature.onObjectEnteredAware(add)
		}
		assert(!creature.isLoggedInPlayer || aware.contains(creature.getSlottedObject("ghost"))) { "not aware of ghost $creature" }
		assert(!creature.isLoggedInPlayer || aware.contains(creature)) { "not aware of creature" }
	}
	
	private inline fun handleFlushDestroy(newAware: Set<SWGObject>, destroyHandler: (Collection<SWGObject>) -> Unit) {
		val destroy = flushAwarenessData.buildDestroy(aware, newAware)
		// Remove destroyed objects immediately - before sending anything to the client
		val it = aware.iterator()
		while (it.hasNext()) {
			val currentAware = it.next()
			for (remove in destroy) { // Since the "create" is filtered, aware could also have been filtered
				if (currentAware === remove || isParent(currentAware, remove)) {
					it.remove()
					awareIds.remove(currentAware.objectId)
					remove.removeObserver(creature)
					creature.onObjectExitedAware(remove)
					break
				}
			}
		}
		destroyHandler(destroy)
		assert(!creature.isLoggedInPlayer || aware.contains(creature.getSlottedObject("ghost"))) { "not aware of ghost $creature" }
		assert(!creature.isLoggedInPlayer || aware.contains(creature)) { "not aware of creature" }
	}
	
	@Synchronized
	fun resetObjectsAware() {
		for (obj in aware) {
			obj.removeObserver(creature)
		}
		aware.clear()
		awareIds.clear()
	}
	
	@Synchronized fun isAware(objectId: Long) = awareIds.contains(objectId)
	@Synchronized fun isAware(obj: SWGObject) = aware.contains(obj)
	
	private fun createObject(obj: SWGObject, target: Player) {
		val id = obj.objectId
		
		// SceneCreateObjectByCrc
		val create = SceneCreateObjectByCrc()
		create.objectId = id
		create.location = obj.location
		create.objectCrc = obj.crc
		target.sendPacket(create)
		
		// Baselines
		val owner = obj.owner === target
		
		target.sendPacket(obj.createBaseline3(target))
		target.sendPacket(obj.createBaseline6(target))
		
		if (owner) {
			target.sendPacket(obj.createBaseline1(target))
			target.sendPacket(obj.createBaseline4(target))
			target.sendPacket(obj.createBaseline8(target))
			target.sendPacket(obj.createBaseline9(target))
		}
		
		// Miscellaneous
		if (obj is CellObject)
			target.sendPacket(UpdateCellPermissionMessage(1.toByte(), id))
		// ? UpdatePostureMessage for PlayerObject ?
		if (obj is CreatureObject && obj.isGenerated()) {
			target.sendPacket(UpdatePostureMessage(obj.posture.id.toInt(), id))
			target.sendPacket(UpdatePvpStatusMessage(obj.pvpFaction, id, this.creature.getPvpFlagsFor(obj)))
		}
		
		// UpdateContainmentMessage
		val parent = obj.parent
		if (parent != null)
			target.sendPacket(UpdateContainmentMessage(obj.objectId, parent.objectId, obj.slotArrangement))
	}
	
	private fun popStackAll(target: Player, createStack: LinkedList<SWGObject>) {
		var parent: SWGObject? = createStack.pollLast()
		while (parent != null) {
			target.sendPacket(SceneEndBaselines(parent.objectId))
			parent = createStack.pollLast()
		}
	}
	
	private fun popStackUntil(target: Player, createStack: LinkedList<SWGObject>, parent: SWGObject) {
		var last: SWGObject? = createStack.peekLast()
		val grandparent = parent.parent
		while (last != null && !(last === parent) && !(last === grandparent)) {
			createStack.pollLast()
			target.sendPacket(SceneEndBaselines(last.objectId))
			last = createStack.peekLast()
		}
	}
	
	private fun destroyObject(obj: SWGObject, target: Player) {
		target.sendPacket(SceneDestroyObject(obj.objectId))
	}
	
	class FlushAwarenessData(private val creature: CreatureObject) {
		
		private val create: MutableCollection<SWGObject> = if (creature is AIObject) ArrayList() else TreeSet(kotlin.Comparator(::compare))
		private val destroy: MutableCollection<SWGObject> = ArrayList()
		
		fun buildCreate(oldAware: Set<SWGObject>, newAware: Set<SWGObject>): Collection<SWGObject> {
			val create = this.create
			create.clear()
			for (obj in newAware) {
				if (!oldAware.contains(obj)) {
					create.add(obj)
				}
			}
			return create
		}
		
		fun buildDestroy(oldAware: Set<SWGObject>, newAware: Set<SWGObject>): Collection<SWGObject> {
			val destroy = this.destroy
			destroy.clear()
			oldAware.asSequence()
					.filter { !newAware.contains(it) }       // Removed from old to new awareness
					.filter { !isParent(creature, it) }      // Only remove if it isn't our parent
					.sortedBy { getObjectDepth(it) }            // Sorted for proper ordering
					.forEach {
						var parent = it.parent
						while (parent != null) {
							if (destroy.contains(parent))
								return@forEach               // Optimization for the client - destroying the parent destroys the children
							parent = parent.parent
						}
						destroy.add(it)
					}
			
			return destroy
		}
		
		private fun compare(a: SWGObject, b: SWGObject): Int {
			var tmpA: SWGObject = a
			var tmpAP: SWGObject? = a.parent
			while (tmpAP != null) {
				if (tmpAP === b)           // A is a child of B
					return 1
				var tmpB: SWGObject = b
				var tmpBP: SWGObject? = b.parent
				while (tmpBP != null) {
					if (tmpBP === tmpA)    // B is a child of A
						return -1
					if (tmpAP === tmpBP) { // Found a shared parent, compare their next highest children
						val bundledA = tmpA.isBundledWithin(tmpAP, creature)
						val bundledB = tmpB.isBundledWithin(tmpBP, creature)
						if (bundledA && !bundledB)
							return -1
						if (bundledB && !bundledA)
							return 1
						return tmpA.objectId.compareTo(tmpB.objectId) // If they're both bundled/not bundled, doesn't matter what order they're in
					}
					tmpB = tmpBP
					tmpBP = tmpB.parent
				}
				tmpA = tmpAP
				tmpAP = tmpA.parent
				if (tmpAP == null) {
					val distComp = getDistance(creature, tmpA).compareTo(getDistance(creature, tmpB))
					if (distComp != 0)
						return distComp
					return tmpA.objectId.compareTo(tmpB.objectId)
				}
			}
			// A is a top level object, since the above loop did not run
			var tmpB: SWGObject = b
			var tmpBP: SWGObject? = b.parent
			while (tmpBP != null) {
				tmpB = tmpBP
				tmpBP = tmpB.parent
			}
			if (tmpA === tmpB)
				return -1 // B is a child of A
			val distComp = getDistance(creature, tmpA).compareTo(getDistance(creature, tmpB))
			if (distComp != 0)
				return distComp
			return tmpA.objectId.compareTo(tmpB.objectId)
		}
		
	}
	
	companion object {
		
		private fun getDistance(creature: CreatureObject, obj: SWGObject) = if (obj.parent != null) 0 else creature.worldLocation.distanceTo(obj.location).toInt()
		private fun SWGObject.isBundledWithin(parent: SWGObject, creature: CreatureObject) = (this.slotArrangement == -1 || this.baselineType == BaselineType.PLAY || parent === creature)
		
		private fun getObjectDepth(obj: SWGObject?): Int {
			var depth = 0
			var tmp = obj
			while (tmp != null) {
				depth++
				tmp = tmp.parent
			}
			return depth
		}
		
		private fun isParent(child: SWGObject, testParent: SWGObject): Boolean {
			var parent: SWGObject? = child.parent
			while (parent != null) {
				if (parent === testParent)
					return true
				parent = parent.parent
			}
			return false
		}
		
	}
	
}
